### PR TITLE
Adds option.ReadTimeout

### PR DESCRIPTION
**Pull request by danielwertheim**
Solves #222 by adding support for configuring `option.ReadTimeout`.

### CODE REVIEW

File: src/Tests/IntegrationTests/TestConnection.cs [hunk 1]
```diff
@@ -453,6 +453,93 @@ public void TestInfineReconnect()
...
+            var payload = new byte[0];
+
+            using (var s = NATSServer.CreateFastAndVerify(Context.Server1.Port))
+            {
+                using (var cn = Context.ConnectionFactory.CreateConnection(opts))
+                {
+                    cn.SubscribeAsync(subject, (sender, args) =>
+                    {
+                        msgEv.Set();
+                        Thread.Sleep(opts.ReadTimeout.Value + 250); //Causes time to pass so that ReadTimeout occurs
```
Comment (watfordgnf) [inline src/Tests/IntegrationTests/TestConnection.cs]: If I'm understanding this test correctly, with the Read Timeout set, any busywork on the message handler can cause this to occur?

If that is the case, should we provide guidance in the `ReadTimeout` documentation?

Comment (danielwertheim) [reply src/Tests/IntegrationTests/TestConnection.cs]: Any suggestion on what you would like the docs to be?

What we do is to affect the `TcpClient.ReceiveTimeout` which in turns sets the underlying `stream.ReadTimeout`.

https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.tcpclient.receivetimeout?view=netframework-4.8#System_Net_Sockets_TcpClient_ReceiveTimeout

I was a bit surprised that I could make "time pass" in that sense in the test, but it's due to the fact that there's nothing inbound during that time as no new publish is made and no ping.

**Comment by danielwertheim** [comment]
We really need to take a look at some flaky tests... Successful build manually provided here (as Azure DevOps has a known issue with re-triggering builds) https://dev.azure.com/NATS-CI/NATS-CI/_build/results?buildId=99&view=results

File: src/Tests/IntegrationTests/TestConnection.cs (hunk 1 above)
Comment (watfordgnf) [reply src/Tests/IntegrationTests/TestConnection.cs]: I'm surprised as well, I guess we'd want remarks letting folks know they should keep all processing to a minimum in their async handlers.

Does this affect using a synchronous message loop?

Comment (danielwertheim) [reply src/Tests/IntegrationTests/TestConnection.cs]: I see this option as something "advanced users" might want to tweak. And in general all handlers should be "short", regardless of this PR or not. Otherwise they still risk getting cut of as a slow consumer, right? Given this, what is there more to change in this PR to complete it?

**Comment by danielwertheim** [comment]
Will that not be a bit confusing as we already have a timeout for `Connect` https://github.com/nats-io/nats.net/blob/master/src/NATS.Client/Conn.cs#L447

I followed the "pattern" that is in place for `processExpectedInfo` https://github.com/nats-io/nats.net/blob/master/src/NATS.Client/Conn.cs#L1194

Setting and reseting the `conn.ReceiveTimeout` during the handshake.

Could do something similar as we do with the send timeout. E.g.

```csharp
if (opts.HandshakeTimeout.HasValue)
{
    var completedInTime = Task.Run(() =>
    {
        processExpectedInfo();
        sendConnect();
    }).Wait(opts.HandshakeTimeout.Value);
    if(!completedInTime)
        throw new NATSConnectionException("Timeout during server handshake.");
}
else
{
    processExpectedInfo();
    sendConnect();
}
```

**Comment by ColinSullivan1** [comment]
@danielwertheim , good point, that would be confusing.... that pattern looks good.

**Comment by ColinSullivan1** [comment]
With the proposed changes, I think the current `Options.Timeout` property would suffice, and we could avoid adding two new properties.  It wouldn't be 100% accurate if this edge case is hit, but close enough imo.  wdyt?

**Comment by danielwertheim** [comment]
But the Timeout is used on the outer scope, right? Thought the idea was to be able to tweak the read side in the connect?

**Comment by ColinSullivan1** [comment]
For most projects, yeah, that'd be a nice to have, but with NATS I'd really like to keep the API footprint as small as possible.  The original issue creator just wanted a timeout, not necessarily to specifically set it... imo just reusing the Timeout is reasonable considering this is a fairly extreme corner case.  I'd be worried users without a lot of networking background could get confused if there are too many timeouts.

**Comment by danielwertheim** [comment]
@ColinSullivan1 Fixed.

File: src/Tests/IntegrationTests/TestConnection.cs (hunk 1 above)
Comment (danielwertheim) [reply src/Tests/IntegrationTests/TestConnection.cs]: Not relevant anymore

**Review by ColinSullivan1 — APPROVED** [review]
LGTM - Thanks!
